### ISSUE TITLE
Under Qubes 4.0, in diskpart.ps1, serialnumber may not match QM00002

### ISSUE THREAD

**Issue by brendanhoar** [tabit-pro/qubes-windows-tools-cross#4]
I believe due to the version of the QEMU build in the stubdomain, QM00002 only matches under Qubes R4.1, not under R4.0.

This prevents usage of the ISO under Qubes R4.0 for building templates.

Brendan

**Comment by jevank** [comment]
We don't test it with 4.0, but even with 4.1 it might be dangerous action, so conditions needs to be improved.

**Comment by brendanhoar** [comment]
I know powershell is a bit limited for verifying it is safe to overwrite the disk, but perhaps we can add additional checks in powershell to ensure we aren't writing over a good disk?

Example code, pulled from a few places...
```
#Utilize pinvoke to call win32 API to get raw disk
Add-Type @"
using System;
using System.Runtime.InteropServices;
public class PInvoke {
    [DllImport("kernel32.dll")] public static extern IntPtr CreateFile(string lpFileName, int dwDesiredAccess, int dwShareMode, IntPtr lpSecurityAttributes, int dwCreationDisposition, int dwFlagsAndAttributes, IntPtr hTemplateFile);
}
"@

$blocksize=512 # assuming 512-size blocks
$blockcount=100 # how many blocks we want to check 
$bytecount=$blocksize*$blockcount # number of bytes to check at start of disk to "prove" disk is empty.

$handle = [PInvoke]::CreateFile('\\.\PhysicalDrive1', 0x80000000, 6, 0, 3, 0, 0)
$fs = New-Object System.IO.FileStream ($handle, 'Read', $true)
$b = [array]::CreateInstance([byte], $bytecount)
$fs.Read($b, 0, $bytecount) | Out-Null
$fs.Dispose()

#Write-Output $b.length

$pass = 1

foreach($byte in $b)
{
	if ($byte -ne 0)
	{
		$pass=0
	}
}

if ($pass -ne 0)
{
	Write-Output 'PHYSICALDISK1 is empty and diskpart can be used to create Private Image.'
}
else
{
	Write-Output 'PHYSICALDISK1 is not empty'
}
#Write data to file
#Set-Content .\disk_header $b -Encoding Byte
```

**Comment by brendanhoar** [comment]
Alternate No. 2: a similar approach in powershell could be used to look for an expected magic number at some location on the disk? The downside is that qubes dom0 tooling would then be expected to mark each of the qubes-created virtual disks to mark them with that number. That way we could be sure (in powershell) that the right disk is being formatted.

**Comment by jevank** [comment]
> Example code, pulled from a few places...

I like the idea with checking zeroes. Did you try this with windows 7 and R4.0?

**Comment by brendanhoar** [comment]
I've tested a modified diskpart.ps1 script under windows 7 and R4.0 and it works: https://github.com/brendanhoar/qubes-windows-tools-cross/blob/Get-diskpart-working-under-R4.0-too/diskpart.ps1

Let me complete testing, via a full ISO build and if it passes for both Win7 and Win10, I'll do a PR.

_Update based on a new install:_

1. When I extracted the ISO from the custom-built RPM based on my branch, I had to use rpm --nosignature to override the following error [package qubes-windows-tools-4.1-65.noarch does not verify: no signature]. I've updated the README.md with a note on that.
2. When executed from the installer the powershell script failed to format the drive.
3. After some debugging I found that on this new install, the new volume was all zeros *except for* two 0x4b (75) bytes in the upper half of the first block on disk. That is very very strange, no idea where they come from.
4. I added a heuristic for determining if the first 32KB of the disk is blank "enough": if the count of unexpected bytes (that is, other than 0x00 and 0x4B) is > 5, do not format. Otherwise, assume disk is empty and format.

Update 2:
With the above, appears stable with Windows 7 under Qubes 4.0 and helpful for getting back to being able to use templating. This change works with Windows 10 as well, which can now be templated under R4.0. I do find Windows 10 less stable with these drivers (maybe only under R4.0?).

Brendan
